### PR TITLE
Finalized script for step 13 - tree cropping

### PR DESCRIPTION
**Pull request by amrithasp02**
This is the working version of tree cropping script I'm using to save all individual tree images. It includes the latest commits made to this file by @b-pardi  in #29 and my changes on top of that. The major change I made was to use the `IDs_to_labels.json` file for each dataset to correctly map the mask instance value to the tree `unique_id` which references the ground truth species labels. I thought it would be good to finalize this and merge since the bigger PR with the model training code is still in progress.

Note: I renamed this file to step 13 since we have `12_render_instance_ids.py`

### CODE REVIEW

File: 1_data_prep/13_tree_crops.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        fp = Path(
+            output_path,                        # base dir to the training data folder
+            species_dir_label,              # separate trees with species labels or no species labels
+            src_info['dset_name'],          # original dataset name (src folder name of images/masks)
+            f"treeID{tree_unique_id}"          # tree id from mask
+        ).resolve()
+        os.makedirs(fp, exist_ok=True)
+
+        # cropped img file name
+        fn = f"treeID{tree_unique_id}-species{species_file_label}-dset{src_info['dset_name']}-view{src_info['flight_type']}.png"
```
[the remark this answers is not in the document]
Comment (russelldj) [reply 1_data_prep/13_tree_crops.py:208]: Yeah, maybe we refactor if/when we do a version 2 of the tree detection and registration work.

File: 1_data_prep/13_tree_crops.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+        # map mask value -> tree unique ID
+        tree_unique_id = ID_mapping.get(int(tree_mask_value))
+        if tree_unique_id is None:
+            raise ValueError(f"Tree unique_ID for mask value {tree_mask_value} is None. Check {IDs_to_labels_path}")
+
+        tree_unique_id = str(int(tree_unique_id)).zfill(5) # convert tree uid to 0 padded str to match plot_attributes format
+
+        # see if this tree has a species label
+        species_val = plot_attributes.loc[plot_attributes['unique_ID'] == tree_unique_id, 'species_code']
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:150]: I think this is where the remap has to happen

File: 1_data_prep/13_tree_crops.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        crosswalk_path (str): Path to the crosswalk CSV file
+        level (str): Mapping level ("l1", "l2", "l3", "l4") or None for no mapping
+    
+    Returns:
+        dict: Mapping from original species_code to target level species code
+        None: If level is None (no mapping)
+    """
+    if level is None:
+        print("No species mapping specified - using original species codes")
+        return None
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:45]: nit: I haven't seen how this is used later, but I could imagine the identity mapping could potentially be returned here and make things simpler downstream.

File: 1_data_prep/13_tree_crops.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+                # No species label
+                final_species = None
+                should_save_crop = True
+                is_mapped = False
+
+            # Determine file naming and directory structure
+            species_file_label = 'NONE' if final_species is None else str(final_species)
+            species_dir_label = 'unlabelled' if final_species is None else 'labelled'
+
+            if not should_save_crop:
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:237]: nit: seems like `should_save_crop` is true in all control flow branches where you haven't hit `continue`.

File: 1_data_prep/13_tree_crops.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            # cropping in PIL is left, top, right, bottom
+            bbox_pad_width, bbox_pad_height = (x1 - x0)  * BBOX_PADDING_RATIO, (y1 - y0) * BBOX_PADDING_RATIO
+            bbox = (x0 - bbox_pad_width, y0 - bbox_pad_height, x1 + bbox_pad_width, y1 + bbox_pad_height)
+            cropped_img = img.crop(bbox)
+
+            fp = Path(
+                output_path,                        # base dir to the training data folder
+                species_dir_label,                  # separate trees with species labels or no species labels
+                src_info['dset_name'],              # original dataset name (src folder name of images/masks)
+                f"treeID{tree_unique_id}"           # tree id from mask
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:279]: nice change

File: 1_data_prep/13_tree_crops.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+            # generate random 10-digit ID for the image
+            image_id = str(uuid.uuid4().int)[:10]   # take first 10 digits of a UUID-derived int
+            crop_path = fp / f"{image_id}.png"
+
+            cropped_img.save(crop_path) # save cropped img
+            mapping_stats['saved_crops'] += 1
+
+            # Build metadata record with dynamic columns based on mapping level
+            record = {
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:291]: I wonder if it would be helpful to include the `img_file_path` here for debugging?

File: 1_data_prep/13_tree_crops.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+for dset_name in dset_names:
+    # metadata records for this specific dataset
+    records = []
+    
+    dset_idx, nadir_id, oblique_id = dset_name.split("_")
+
+    # Check if output folder for this dataset already exists in 'labelled'
+    labelled_output_dir = Path(output_path, "labelled", dset_name)
+    if labelled_output_dir.exists():
```
Comment (russelldj) [inline 1_data_prep/13_tree_crops.py:117]: Can we check that the metadata file has been created, rather than the top level folder? That would ensure all the crops have been generated, which could be helpful if it dies in the middle of a given dataset.

**Review by russelldj — COMMENTED** [review]
Looks very close to me, just comments on the new content.

File: 1_data_prep/13_tree_crops.py (hunk 3 above)
Comment (amrithasp02) [reply 1_data_prep/13_tree_crops.py:45]: The output from this function is used in the method below, called `map_species` and the identity mapping is essentially happening there since it just returns the original species code when `species_mapping = None`

**Review by russelldj — APPROVED** [review]
Great!
